### PR TITLE
Buy Pool - Last Steps

### PR DESCRIPTION
**Pull request by gustavoguimaraes**
I am coming back to this code after two weeks and I am still recalling what we did here. 

I added on top of @elektronaut0815 modifications. Basically added natspec where they were missing and made them more coherent.

Can you look at it so we move forward in adding the buying functionality?

### CODE REVIEW

File: contracts/Pool.sol [hunk 1]
```diff
@@ -21,16 +21,13 @@ contract Pool {
     IEtherToken public token1;
     ERC20 public token2;
 
-//    uint256 fundedBuyValueUSD = 0;
```
Comment (elektronaut0815) [inline contracts/Pool.sol]: Might still need this later.

File: contracts/Pool.sol [hunk 2]
```diff
@@ -170,22 +175,9 @@ contract Pool {
         );
 
         buyContributorToken2Amount[msg.sender] = buyContributorToken2Amount[msg.sender].add(buyWithToken2Amount);
-
-        emit ContributeToBuyPool(msg.sender, address(token2), buyWithToken2Amount);
-
         token2BuyBalance = token2BuyBalance.add(buyWithToken2Amount);
 
-/*        uint256 addedBuyValueUSD = isAuctionWithWeth ?
```
Comment (elektronaut0815) [inline contracts/Pool.sol]: We might still need this.

**Comment by gustavoguimaraes** [comment]
I made a major refactor of the code so currently it does the following:

**contribute()**  triggers `_addTokenPair()` only if the the contribution for token1 matches the sufficient volume contributed for token2 after taking into account the price.

**collectFunds()** calls ducthX `postBuyOrder()` passing the collected token2.

Tests that were already in this branch are passing. I needed to remove some of the open-zeppelin-test-helpers nonsense as it was not working well with ganache blockchain instance within our truffle local setup.

There needs to be added more tests. Especially related if the funds are taken correctly from DutchX back to the pool. 
The pool needs to receive back all the fund (minus dutchX fees?) that it sent to DutchX.

File: contracts/Pool.sol [hunk 3]
```diff
@@ -20,12 +20,15 @@ contract Pool {
     IEtherToken public token1;
     ERC20 public token2;
 
+    uint256 fundedValueUSD;
```
Comment (gustavoguimaraes) [inline contracts/Pool.sol]: Note: I don't think we need to declare this var globally

**Comment by gorgos** [comment]
@gustavoguimaraes Great idea to merge the `contribute` for buy and sell.

- Buy side will not added to the DutchX via `addTokenPair` by passing 0, that should work. ✔️ 
- The threshold `fundedValueUSD` is based on buy and sell combined. This is not correct. We need to compute a threshold for sell and buy. And both sides must be exactly matching. ❌ 

Therefore, I propose the following.

1) Compute sell and buy thresholds separately.
2) Stop sell side pool once threshold is reached. Refund any extra sent tokens to last contributer.
3) Stop buy side pool once threshold is reached. Refund any extra sent tokens to last contributer.
4) Both thresholds reached? -> `addTokenPair`

**Comment by gustavoguimaraes** [comment]
Agree. Let's do It

File: contracts/Pool.sol [hunk 4]
```diff
@@ -220,6 +214,33 @@ contract Pool {
...
+        stage = Stages.Claim;
+        uint256 auctionIndex = dx.getAuctionIndex(address(token1), address(token2));
+
+        // post buy order
+        token2.approve(address(dx), token2Balance);
+        dx.deposit(address(token2), token2Balance);
+        dx.postBuyOrder(address(token1), address(token2), auctionIndex, token2Balance);
+
+        // claim funds to pool
+        dx.claimSellerFunds(address(token1), address(token2), address(this), 1);
```
Comment (gustavoguimaraes) [inline contracts/Pool.sol]: Note: one thing I am not certain is how the claim for the buy side happens.

Here we claim the sellerFunds

**Comment by gustavoguimaraes** [comment]
@elektronaut0815 looking good

File: contracts/Pool.sol (hunk 4 above)
Comment (gorgos) [reply contracts/Pool.sol]: Good point, we are missing this!

https://github.com/gnosis/dx-contracts/blob/31e43f19762247d4b4e945f52c08a9a3ef2ea264/contracts/DutchExchange.sol#L554

`claimBuyerFunds` !

\cc @loki-sama @elektronaut0815

**Pull request merged**